subsys/fs/littlefs_fs.c

```diff
@@ -587,19 +587,23 @@ static int littlefs_mount(struct fs_mount_t *mountp)
 	fs_lock(fs);
 
 	/* Open flash area */
-	ret = flash_area_open(area_id, &fs->area);
-	if ((ret < 0) || (fs->area == NULL)) {
+	const struct flash_area **fap = (const struct flash_area **)&fs->backend;
+
+	ret = flash_area_open(area_id, fap);
```

Comment:
I did not get it, sorry. flash_area_open() sets **fa based on id (btw note to maintainers, it does not set it to NULL on error/id unknown, check flash_area_open() implementation). Above **fap is initialized to &fs->backend, what is when `fs->backend = NULL`?, then used as parameter for flash_area_open() and line below fs->backend is checked for NULL, and all these casts...

Why not 

```c
        const struct flash_area **fap = NULL;  

...

        ret = flash_area_open(area_id, fap);                                                                                                 
        if ((ret < 0) || (fap == NULL)) {                                                                                                    
                LOG_ERR("Can't open flash area %d", area_id);                                                                                
                return -ENODEV;                                                                                                              
        }                                                                                                                                    
                                                                                                                                             
        LOG_DBG("FS area %u at 0x%x for %u bytes",                                                                                           
                area_id, (uint32_t)(*fap)->fa_off,                                                                                           
                (uint32_t)(*fap)->fa_size);                                                                                                  
                                                                                                                                             
        dev = flash_area_get_device(*fap);                                                                                                   
        if (dev == NULL) {                                                                                                                   
                LOG_ERR("Can't get flash device: %s",                                                                                        
                        log_strdup((*fap)->fa_dev_name));                                                                                    
                return -ENODEV;                                                                                                              
        }                                                                                                                                    
                                                                                                                                             
        fs->backend = fap;
```
?

Reply:
Casting was used to keep the consistency for `fs->backend` - it can be cast'ed to `(char *)` on the block device. Or to `const struct flash_area *` when flash device is used.

However, as the `littlefs_flash_init()` is a small function, the usage of `**fap` pointer is better solution than the one proposed by me. I will adjust this function.

Reply:
Checking `fap == NULL` after you got `ret < 0` should not be required as `fap` should always be not `NULL` if `ret >= 0`

Reply:
> Actually it only does not set it to NULL when returning -EACCS, which happens when there is no flash map defined at all. I am not sure whether this is bug.

It is not a bug as the exit of the routine will set fs->area = NULL.

Reply:
> > Actually it only does not set it to NULL when returning -EACCS, which happens when there is no flash map defined at all. I am not sure whether this is bug.
> 
> It is not a bug as the exit of the routine will set fs->area = NULL.

Not in the -EACCS exit path, when there is not map at all. But it does not matter you are supposed to check ret anyway.

Reply:
offtopic but...
_Pointer which has to reference flash_area. If @p ID is unknown, it will be NULL on output._, see https://github.com/zephyrproject-rtos/zephyr/blob/main/include/storage/flash_map.h#L117

Implementation:
```c
int flash_area_open(uint8_t id, const struct flash_area **fap)                                                                               
{                                                                                                                                            
        const struct flash_area *area;                                                                                                       
                                                                                                                                             
        if (flash_map == NULL) {                                                                                                             
                return -EACCES;                                                                                                              
        }                                                                                                                                    
                                                                                                                                             
        area = get_flash_area_from_id(id);                                                                                                   
        if (area == NULL) {                                                                                                                  
                return -ENOENT;                                                                                                              
        }                                                                                                                                    
                                                                                                                                             
        *fap = area;                                                                                                                         
        return 0;                                                                                                                            
} 
```

get_flash_area_from_id(id) returns either pointer to flash_area or NULL, fap assigned only if `area != NULL`, no issue here? That also why compiler complains if fap is passed to flash_area_open() uninitialized.

Reply:
If we take this literally then we do not know whether ID is known because we have no map. But, yeah seems that this is bug, although I ma not sure whether fap should get modified at all when bug occurs. Log an issue for this.

Reply:
https://github.com/zephyrproject-rtos/zephyr/pull/41723

Reply:
Not addressed

Reply:
@lmajewski, a cleaner method to avoid casting in the flash_init:
```
static int littlefs_flash_init(struct fs_mount_t *mountp)
{
	unsigned int area_id = (uintptr_t)mountp->storage_dev;
	struct fs_littlefs *fs = mountp->fs_data;
        struct flash_area *area;
	const struct device *dev;
        int ret;

	ret = flash_area_open(area_id, &area);
	if (ret < 0) {
		LOG_ERR("can't open flash area %d", area_id);
		return -ENODEV;
	}

	LOG_DBG("FS area %u at 0x%x for %u bytes", area_id,
                (uint32_t)(area->fa_off,(uint32_t)(area->fa_size);

	dev = flash_area_get_device(area);
	if (dev == NULL) {
		LOG_ERR("can't get flash device: %s", 
                        log_strdup(area->fa_dev_name));
		return -ENODEV;
	}

        fs->device = (void *)area;
	return 0;
}
```

Reply:
@Laczen Thanks for the code snippet, but I've followed the idea propsed by @jfischer-no with the `**fap` pointer.

Reply:
That is OK for now, when I do cleaning up I will change to my proposal as this is how flash_area_open() is used in zephyr.